[IP-763]: feature: Credit Note/Invoice

Verony-makesIT opened the pull request:
## Description
<!--
Please check the following steps to submit your pull request. If you have any questions please read the contribution guide available at https://go.invoiceplane.com/contribguide or join the community forums or the Slack channel.
You can check items by changing `[ ]` to `[x]`.
If you can't check all checklist items please add `[WIP]` in front of your title.
Remove this first paragraph but please keep the following checklist even if it's incomplete.
-->

## Related Issue
1. #763
2. [Credit invoice date not correct](https://community.invoiceplane.com/t/topic/10660)
3. [Credit invoice template feature developed](https://community.invoiceplane.com/t/topic/10644)

## Motivation and Context
This PR resolves the issue (Related Issue 2.). 
Also creates an additional Invoice group item for the purpose of credit notes.
When creating a credit note, it will: 
- change all invoice entries to credit note entries including credit note terms in the invoice form and templates. (Related Issue 1.)
- change the creation date to today's date. (Related Issue 2.) 
- check whether the 'Default Credit Note/Invoice Group' in Settings has already been filled in.

## Screenshots (if appropriate):

## Pull Request Checklist

  * [X] My code follows the code formatting guidelines.
  * [X] I have an issue ID for this pull request.
  * [X] I selected the corresponding branch.
  * [X] I have rebased my changes on top of the corresponding branch.

## Issue Type (Please check one or more)

  * [X] Bugfix
  * [X] Improvement of an existing Feature
  * [ ] New Feature

Verony-makesIT commented:
Here you can find some more relevant screenshots. 

Credit Invoice page:
![image](https://github.com/InvoicePlane/InvoicePlane/assets/30659226/16d2f769-2df0-46f5-b466-4b48ad199034)

Settings - Invoice Group Form & Invoices Group and Terms 
![image](https://github.com/InvoicePlane/InvoicePlane/assets/30659226/c256c99d-dce1-49fd-9261-baa5d8cb7bc3)

Modal (Credit) Invoice Group: not selected!
![image](https://github.com/InvoicePlane/InvoicePlane/assets/30659226/181369ab-6f83-4716-a416-585dc4977060)

InvoicePlane PDF
![image](https://github.com/InvoicePlane/InvoicePlane/assets/30659226/58e22507-7376-4016-8ee0-a146accb38c2)

InvoicePlane_Web:
![image](https://github.com/InvoicePlane/InvoicePlane/assets/30659226/999c8755-b45b-4aa8-9f62-f1040d26b054)

aksysict commented:
Using qr code and the invoice_sign is -1 it gives an error, there needs to be a check before loading qr code.

aksysict commented:
> @aksysict, thanks for reporting this. I will look into this ASAP.

with the code you added for the qr code, you get a empty pdf only with logo en company and client details no details about the items and prices

Verony-makesIT commented:
@aksysict,
I added a check only for the QR-Code (if invoice_sign != -1) to the pdf template.
Then I created a new invoice and subsequently created a credit invoice from it (**_without any problems!_**).
See this screenshot of the invoice and of the credit invoice pdf (**including items and prices**).
![image](https://github.com/InvoicePlane/InvoicePlane/assets/30659226/51e687fb-ea44-427d-9cec-ca8b53ee6d76)

So this should no longer cause a problem.

naui95 commented on application/modules/invoices/controllers/Ajax.php:
```diff
@@ -435,9 +435,16 @@ public function create_credit()
             $this->mdl_invoices->where('invoice_id', $target_id);
             $this->mdl_invoices->update('ip_invoice_amounts', ['invoice_sign' => '-1']);
 
+            // Update the invoice date created (today) and due date
```
Is it right for a credit note to have a due date? The function is also setting a due date.

Verony-makesIT replied:
_I know..._
I see a **'credit note due date'** in IP dashboard as an indication that I still need to take action and I find it useful because I can immediately see that there is still an (_expired?_) refund to be made to my customer.
As a result, the (_credit note_) due date in the **InvoicePlane template** must of course also be adjusted.

naui95 commented on application/modules/settings/views/partial_settings_invoices.php:
```diff
@@ -81,6 +81,38 @@ class="form-control" value="<?php echo get_setting('invoices_due_after'); ?>">
+                                <?php _trans('default_creditnote_terms'); ?>
+                            </label>
+                            <textarea name="settings[default_creditnote_terms]" id="settings[default_creditnote_terms]"
+                                class="form-control" rows="1"><?php echo get_setting('default_creditnote_terms', '', true); ?>
```
Please use 4 rows as for the invoice notes.

Verony-makesIT replied:
Changed <textarea> to 4 rows.

Verony-makesIT commented on application/modules/invoices/views/view.php:
```diff
@@ -376,7 +381,13 @@ function DownR(k) {
-                                    <label><?php _trans('invoice'); ?> #</label>
+                                    <label>
+                                        <?php if ($invoice->invoice_sign == -1) {
+                                                echo _trans('credit_invoice');
```
Changed to echo trans().

Verony-makesIT commented on application/modules/invoices/views/modal_create_credit.php:
```diff
@@ -62,7 +62,7 @@ function (data) {
                     <?php foreach ($invoice_groups as $invoice_group) { ?>
                         <option value="<?php echo $invoice_group->invoice_group_id; ?>"
-                            <?php if (get_setting('default_invoice_group') == $invoice_group->invoice_group_id) {
+                            <?php if (get_setting('default_creditnote_group') == $invoice_group->invoice_group_id) {
```
As requested in FR #763 (separate template and CNI number) and also because the current procedure in my country is not correct, I have created this PR. 
If I have to apply your proposal, invoice numbers and Credit Note numbers are mixed/used. 
See screenshot.
![image](https://github.com/user-attachments/assets/137a0f13-1d59-47f3-8e72-0c893f8ea0bb)
![image](https://github.com/user-attachments/assets/5e088bf8-0774-4ca9-8bde-176ac94a40bf)

I propose to implement a (credit note) choice or switch (**use invoice numbering or separate credit note numbering**) in "System Settings - Invoices" so that the user can decide for themselves which type they want to use.

Verony-makesIT commented:
@naui95,  "**This branch has conflicts that must be resolved**"
As stated here I'm going to remove the 6 commits and then match the branch to the the upstream repository. 
Then I'm going to resubmit the commits.

Verony-makesIT commented:
@naui95 
Apparently Github has automatically closed my PR.
How can I reopen it so I can continue making changes?
Any advice is welcome...

nielsdrost7 commented:
The button to reopen the PR is disabled.
Your own branch and the branch you want to merge in to are exactly the same.

@Verony-makesIT you have 2 options to reopen the PR:
1. add extra "source" to your own branch, maybe an "enter" somewhere, or a translation.
That way there's a difference between your own branch and the branch you want to merge in to and then we can reopen the PR.

2. Personally I would advice to add your improvements to the branch on the InvoicePlane repository

Verony-makesIT commented:
@nielsdrost7,
_I didn't know what else to do and just set up a new branch "**CRedit-Invoice-(CRI)-template**" ..._
**Should I create a new PR now or reopen this one...?**
> So I have added a switch for credit invoices (sales/credit) with different numbering and different layout in "System Settings - Invoices".
> I have also added a separate "Credit Invoice terms" item so that it can be used/set by both templates.
> By default it is set to the invoice number template (sales invoices).
> Activate the option "Enable a different Credit Invoice format" and select the correct "Credit Invoice Group" to use a different template than the "Invoice Default".

![image](https://github.com/user-attachments/assets/9a336bba-7181-42a6-a32a-be7c25800f5e)

![image](https://github.com/user-attachments/assets/09fb2791-9d22-4338-84a3-9528e7fafa78)

nielsdrost7 commented:
@Verony-makesIT could you open a new PR please.
Just copy/paste your old text in that new PR, the text was ok

Verony-makesIT commented:
@nielsdrost7,
I have created a new PR [IP-763: feature: Credit Note/Invoice 1116](https://github.com/InvoicePlane/InvoicePlane/pull/1116) and hope that everything is now in order.
There is a (_small_) difference (_**a CRI switch**_) with the original setup of the old PR.
Is it necessary to add the screenshot of my previous comment above (https://github.com/InvoicePlane/InvoicePlane/pull/1026#issuecomment-2474079931) in the new PR?

nielsdrost7 commented:
@Verony-makesIT 
Thank you for making the new PR. I'll take a look this weekend.

> Is it necessary to add the screenshot of my previous comment above (https://github.com/InvoicePlane/InvoicePlane/pull/1026#issuecomment-2474079931) in the new PR?

That's ok. I'll add it when I need the context to see what you created.
